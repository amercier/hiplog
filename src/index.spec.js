import * as hiplog from '.';

/** @test {hiplog} */
describe('hiplog', () => {
  it('exists', () => expect(hiplog).toBeDefined());
  it("doesn't export a default", () => expect(hiplog.default).not.toBeDefined());
});

/** @test {defaultOptions} */
describe('defaultOptions', () => {
  const { defaultOptions } = hiplog;
  it('has 10 members', () => expect(Object.keys(defaultOptions).length).toBe(10));
});

/** @test {Log} */
describe('Log', () => {
  const { Log } = hiplog;
  it('exists', () => expect(Log).toBeDefined());

  describe('constructor', () => {
    const { defaultOptions } = hiplog;

    it('assigns defaults options', () => {
      const log = new Log();
      Object.keys(defaultOptions).forEach(
        key => expect(log.options).toHaveProperty(key, defaultOptions[key]),
      );
    });

    it('assigns given options', () => {
      Object.keys(defaultOptions).forEach((key) => {
        const value = key === 'levels' ? ['foo', 'bar'] : 'foo';
        const log = new Log({ [key]: value });
        expect(log.options).toHaveProperty(key, value);
      });
    });

    it('create one function member for each level', () => {
      const levels = ['foo', 'bar', 'baz'];
      const log = new Log({ levels });
      levels.forEach(level => expect(log[level]).toBeFunction());
    });
  });

  describe('write', () => {
    it('is called with given level by created function members', () => {
      const levels = ['foo', 'bar'];
      const log = new Log({ levels });
      log.write = jest.fn();

      levels.forEach((level, i) => {
        log[level]('baz', 'foobar');
        expect(log.write).toHaveBeenCalledTimes(1);
        expect(log.write).toHaveBeenCalledWith(i, ['baz', 'foobar']);
        log.write.mockClear();
      });
    });

    it('writes whatever is returned by buildMessage() to stream', () => {
      const levels = ['foo'];
      const log = new Log({ levels, level: levels[0] });
      log.buildMessage = () => 'bar';
      log.options.stream = { write: jest.fn() };

      log.foo();
      expect(log.options.stream.write).toHaveBeenCalledTimes(1);
      expect(log.options.stream.write).toHaveBeenCalledWith('bar');
    });

    it("doesn't write anything to stream when level is below message level", () => {
      const log = new Log({ level: 'info' });
      log.buildMessage = level => log.options.levels[level];
      log.options.stream = { write: jest.fn() };

      log.debug();
      log.info();
      expect(log.options.stream.write).toHaveBeenCalledTimes(1);
      expect(log.options.stream.write).toHaveBeenCalledWith('info');
    });
  });

  describe('buildMessage', () => {
    it('builds simple messages', () => {
      const log = new Log({ level: 'debug', displayTime: false });
      const messsages = [
        ['debug', 'messages to debug an application'],
        ['info', 'a purely informational message'],
        ['notice', 'a normal but significant condition'],
        ['warning', 'warning condition'],
        ['error', 'error condition'],
        ['critical', 'the system is in critical condition'],
        ['alert', 'action must be taken immediately'],
        ['emergency', 'system is unusable'],
      ].map(([level, ...parts]) => log.buildMessage(log.options.levels.indexOf(level), parts));
      expect(messsages).toMatchSnapshot();
    });

    it('builds messages with multiple parts', () => {
      const log = new Log({ level: 'debug', displayTime: false });
      const messsages = [
        ['debug', 'messages', 'to', 'debug', 'an', 'application'],
        ['info', 'a', 'purely', 'informational', 'message'],
        ['notice', 'a', 'normal', 'but', 'significant', 'condition'],
        ['warning', 'warning', 'condition'],
        ['error', 'error', 'condition'],
        ['critical', 'the', 'system', 'is', 'in', 'critical', 'condition'],
        ['alert', 'action', 'must', 'be', 'taken', 'immediately'],
        ['emergency', 'system', 'is', 'unusable'],
      ].map(([level, ...parts]) => log.buildMessage(log.options.levels.indexOf(level), parts));
      expect(messsages).toMatchSnapshot();
    });

    it('builds messages with non-string parts', () => {
      const log = new Log({ displayTime: false });
      const level = log.options.levels.indexOf('info');
      const messsages = [
        ['boolean:', true, false],
        ['null:', null],
        ['undefined:', undefined],
        ['number:', 1, 2, 3],
        ['regexp:', /foo/],
        ['array:', ['foo', 'bar']],
        ['object:', { foo: 'bar' }],
      ].map(parts => log.buildMessage(level, parts));
      expect(messsages).toMatchSnapshot();
    });

    it('builds messages with big objects', () => {
      const log = new Log({ displayTime: false });
      const circular = {};
      circular.inner = circular;
      const bigObject = {
        null: null,
        undefined,
        integer: 123,
        boolean: true,
        string: 'Hello',
        funtion: function myFunction() {},
        circular,
        array: ['one', 'two', 'three', 'four'],
      };
      const messsage = log.buildMessage(log.options.levels.indexOf('info'), [bigObject]);
      expect(messsage).toMatchSnapshot();
    });

    it('builds messages with errors', () => {
      const log = new Log({ displayTime: false });
      let error;
      try {
        throw new Error('LEROOOOOOOOOOOY JENKINS!!!!!!!!!!!!!!!!!!!!!!!!!');
      } catch (e) {
        error = e;
      }
      const messsage = log.buildMessage(log.options.levels.indexOf('error'), [error]);
      expect(messsage).toMatchSnapshot();
    });
  });
});
