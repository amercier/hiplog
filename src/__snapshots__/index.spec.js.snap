// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log buildMessage builds messages with big objects 1`] = `
"[34m INFO[39m[34m â€¢ [39m[90m[39m
        [90m{[39m
        [90m  [1m[37mnull[90m[22m: [1m[31mnull[90m[22m,[39m
        [90m  [1m[37mundefined[90m[22m: [7m[31mundefined[90m[27m,[39m
        [90m  [1m[37minteger[90m[22m: [1m[34m123[90m[22m,[39m
        [90m  [1m[37mboolean[90m[22m: [1m[32mtrue[90m[22m,[39m
        [90m  [1m[37mstring[90m[22m: [33m'Hello'[90m,[39m
        [90m  [1m[37mfuntion[90m[22m: [36m[Function: myFunction][90m,[39m
        [90m  [1m[37mcircular[90m[22m: {[39m
        [90m    [1m[37minner[90m[22m: [1m[90m[Circular~ circular][90m[22m[39m
        [90m  },[39m
        [90m  [1m[37marray[90m[22m: [[39m
        [90m    [33m'one'[90m,[39m
        [90m    [33m'two'[90m,[39m
        [90m    [33m'three'[90m,[39m
        [90m    [33m'four'[90m[39m
        [90m  ][39m
        [90m}[39m
        [90m[39m
"
`;

exports[`Log buildMessage builds messages with errors 1`] = `
"[31mERROR[39m[31m â€¢ [39m[31mLEROOOOOOOOOOOY JENKINS!!!!!!!!!!!!!!!!!!!!!!!!![39m
        
        [0m [90m 170 | [39m      let error[33m;[39m[0m
        [0m [90m 171 | [39m      [36mtry[39m {[0m
        [0m[31m[1m>[22m[39m[90m 172 | [39m        [36mthrow[39m [36mnew[39m [33mError[39m([32m'LEROOOOOOOOOOOY JENKINS!!!!!!!!!!!!!!!!!!!!!!!!!'[39m)[33m;[39m[0m
        [0m [90m     | [39m              [31m[1m^[22m[39m[0m
        [0m [90m 173 | [39m      } [36mcatch[39m (e) {[0m
        [0m [90m 174 | [39m        error [33m=[39m e[33m;[39m[0m
        [0m [90m 175 | [39m      }[0m
        
          [2mat Object.<anonymous> ([22mhiplog/src/index.spec.js[2m:172:15)[22m
        
"
`;

exports[`Log buildMessage builds messages with multiple parts 1`] = `
Array [
  "[90mDEBUG[39m[90m â€¢ [39mmessages to debug an application
",
  "[34m INFO[39m[34m â€¢ [39ma purely informational message
",
  "[33m NOTE[39m[33m â€¢ [39ma normal but significant condition
",
  "[35m WARN[39m[35m â€¢ [39mwarning condition
",
  "[31mERROR[39m[31m â€¢ [39merror condition
",
  "[31m[1mCRITI[22m[39m[31m[1m â€¢ [22m[39mthe system is in critical condition
",
  "[31m[7mALERT[27m[39m[31m â€¢ [39maction must be taken immediately
",
  "[31m[1m[7mEMERG[27m[22m[39m[31m[1m â€¢ [22m[39msystem is unusable
",
]
`;

exports[`Log buildMessage builds messages with non-string parts 1`] = `
Array [
  "[34m INFO[39m[34m â€¢ [39mboolean: [90m[1m[32mtrue[90m[22m[39m [90m[1m[31mfalse[90m[22m[39m
",
  "[34m INFO[39m[34m â€¢ [39mnull: [90m[1m[31mnull[90m[22m[39m
",
  "[34m INFO[39m[34m â€¢ [39mundefined: [90m[7m[31mundefined[90m[27m[39m
",
  "[34m INFO[39m[34m â€¢ [39mnumber: [90m[1m[34m1[90m[22m[39m [90m[1m[34m2[90m[22m[39m [90m[1m[34m3[90m[22m[39m
",
  "[34m INFO[39m[34m â€¢ [39mregexp: [90m[35m/foo/[90m[39m
",
  "[34m INFO[39m[34m â€¢ [39marray: [90m[ [33m'foo'[90m, [33m'bar'[90m ][39m
",
  "[34m INFO[39m[34m â€¢ [39mobject: [90m{ [1m[37mfoo[90m[22m: [33m'bar'[90m }[39m
",
]
`;

exports[`Log buildMessage builds simple messages 1`] = `
Array [
  "[90mDEBUG[39m[90m â€¢ [39mmessages to debug an application
",
  "[34m INFO[39m[34m â€¢ [39ma purely informational message
",
  "[33m NOTE[39m[33m â€¢ [39ma normal but significant condition
",
  "[35m WARN[39m[35m â€¢ [39mwarning condition
",
  "[31mERROR[39m[31m â€¢ [39merror condition
",
  "[31m[1mCRITI[22m[39m[31m[1m â€¢ [22m[39mthe system is in critical condition
",
  "[31m[7mALERT[27m[39m[31m â€¢ [39maction must be taken immediately
",
  "[31m[1m[7mEMERG[27m[22m[39m[31m[1m â€¢ [22m[39msystem is unusable
",
]
`;

exports[`defaultOptions has correct values 1`] = `
Object {
  "colors": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "displayTime": false,
  "displayTimeFormat": "yyyy-mm-dd HH:MM:ss.l",
  "format": [Function],
  "inverse": Array [
    true,
    true,
  ],
  "labels": Array [
    "EMERG",
    "ALERT",
    "CRITI",
    "ERROR",
    " WARN",
    " NOTE",
    " INFO",
    "DEBUG",
  ],
  "level": "info",
  "levels": Array [
    "emergency",
    "alert",
    "critical",
    "error",
    "warning",
    "notice",
    "info",
    "debug",
  ],
  "separator": " â€¢ ",
  "stream": [Function],
}
`;

exports[`fromEnv uses development options when \`NODE_ENV\` set to "development" 1`] = `
Object {
  "colors": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "displayTime": false,
  "displayTimeFormat": "yyyy-mm-dd HH:MM:ss.l",
  "format": [Function],
  "inverse": Array [
    true,
    true,
  ],
  "labels": Array [
    "EMERG",
    "ALERT",
    "CRITI",
    "ERROR",
    " WARN",
    " NOTE",
    " INFO",
    "DEBUG",
  ],
  "level": "info",
  "levels": Array [
    "emergency",
    "alert",
    "critical",
    "error",
    "warning",
    "notice",
    "info",
    "debug",
  ],
  "separator": " â€¢ ",
  "stream": [Function],
}
`;

exports[`fromEnv uses test options when \`NODE_ENV\` set to "test" 1`] = `
Object {
  "colors": Array [
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
    [Function],
  ],
  "displayTime": false,
  "displayTimeFormat": "yyyy-mm-dd HH:MM:ss.l",
  "format": [Function],
  "inverse": Array [
    true,
    true,
  ],
  "labels": Array [
    "EMERG",
    "ALERT",
    "CRITI",
    "ERROR",
    " WARN",
    " NOTE",
    " INFO",
    "DEBUG",
  ],
  "level": "critical",
  "levels": Array [
    "emergency",
    "alert",
    "critical",
    "error",
    "warning",
    "notice",
    "info",
    "debug",
  ],
  "separator": " â€¢ ",
  "stream": [Function],
}
`;
